WIP: Add support for dual stack load balancers

nrb opened the pull request:
**What type of PR is this?**
/kind feature


**What this PR does / why we need it**:

Adds basic support for configuring the IP address family on Services.

This functionality overlaps with the ALBC somewhat, but the goal here is to provide much more basic functionality, not the full scope of ALBC.

**Which issue(s) this PR fixes**:
<!-- 
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #1219 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Support for the Kubernetes Service dual-stack networking API.
```

damdo commented:
Hey @kmala do we know what's happening with the Netlify checks? It looks like they are broken across many open PRs. TY

damdo commented:
/retest

nrb commented:
/test pull-cloud-provider-aws-test

One more run to see if this is a rate limiting issue or some sort of genuine issue with my code.

tthvo commented on pkg/providers/v1/aws.go:
```diff
@@ -234,6 +234,11 @@ const ServiceAnnotationLoadBalancerEIPAllocations = "service.beta.kubernetes.io/
+// ServiceAnnotationLoadBalancerIPAddressType is the annotation used on the service
+// to specify the IP address type for the load balancer. Supported values are "ipv4" and "dualstack".
+// Defaults to "ipv4". Only supported on NLB.
+const ServiceAnnotationLoadBalancerIPAddressType = "service.beta.kubernetes.io/aws-load-balancer-ip-address-type"
```
Wondering if we should add support for configuring Target Group IP type (i.e. `ipv4` and `ipv6`)?

Currently, the Target Groups are created as `ipv4` only regardless of the IP type of Load Balancer.

nrb replied:
Yes, we should definitely do this. I was thinking of doing this in a separate PR in order to keep the changes more manageable. Do you think it would be better as a commit within this one?

tthvo reviewed (commented):
I tested with a simple service below:

```yaml
apiVersion: v1
kind: Service
metadata:
  name: wordpress-mysql
  labels:
    app: wordpress
  annotations:
    service.beta.kubernetes.io/aws-load-balancer-type: nlb
    service.beta.kubernetes.io/aws-load-balancer-ip-address-type: dualstack
spec:
  ipFamilyPolicy: PreferDualStack
  ports:
    - port: 3306
  selector:
    app: wordpress
  type: LoadBalancer
```

The CCM does create the dualstack NLB as expected 🚀  

```bash
$ kubectl -n myns get svc/wordpress-mysql -o yaml | yq .status.loadBalancer.ingress
- hostname: <id>.elb.us-east-1.amazonaws.com

$ nslookup <id>.elb.us-east-1.amazonaws.com
# nslookup shows both public IPv4 EIP and IPv6 GUA addresses

$ aws elbv2 describe-load-balancers --names=<nlb-id> --query='LoadBalancers[0].[Type,IpAddressType,Scheme]'
[
    "network",
    "dualstack",
    "internet-facing"
]
```

nrb commented:
/test pull-cloud-provider-aws-test

nrb commented:
/test pull-cloud-provider-aws-check

mtulio commented on pkg/providers/v1/aws.go:
```diff
@@ -234,6 +234,16 @@ const ServiceAnnotationLoadBalancerEIPAllocations = "service.beta.kubernetes.io/
+// to specify the IP address type for the target groups. Supported values are "ipv4" and "ipv6".
+// Defaults to "ipv4". Only supported on NLB.
+const ServiceAnnotationLoadBalancerTargetGroupIPAddressType = "service.beta.kubernetes.io/aws-load-balancer-target-group-ip-address-type"
+
```
Please also update the service controller documentation: https://github.com/kubernetes/cloud-provider-aws/blob/master/docs/service_controller.md

mtulio commented on pkg/providers/v1/aws.go:
```diff
@@ -2483,6 +2512,12 @@ func (c *Cloud) EnsureLoadBalancer(ctx context.Context, clusterName string, apiS
 		if len(sourceRangeCidrs) == 0 {
 			sourceRangeCidrs = append(sourceRangeCidrs, "0.0.0.0/0")
+
+			// For dual-stack or IPv6 load balancers, also add IPv6 default route
```
route or rules?
```suggestion
			// For dual-stack or IPv6 load balancers, also add IPv6 default rules
```

sadasu replied:
Route.  "0.0.0.0/0" and "::/0" act as a gateway of last resort, directing unmatched traffic to a specific next-hop or upstream router.

mtulio commented on pkg/providers/v1/aws_validations.go:
```diff
@@ -98,6 +103,64 @@ func validateServiceAnnotations(v *awsValidationInput) error {
+		switch *v.apiService.Spec.IPFamilyPolicy {
+		case v1.IPFamilyPolicyRequireDualStack:
+			// RequireDualStack is only supported for NLB
+			if !isNLB {
```
this was checked in the outer function, or was that added to increase unit coverage? if so, maybe returning early too?

mtulio reviewed (commented):
Great work! A few nits, and questions related to the restrictions on target ipv6.

tthvo commented on pkg/providers/v1/aws_loadbalancer.go:
```diff
@@ -886,13 +968,48 @@ func (c *Cloud) ensureTargetGroupTargets(ctx context.Context, tgARN string, expe
+		}
+		if len(networkInterface.Ipv6Addresses) > 0 {
+			// Return the first IPv6 address
+			return aws.ToString(networkInterface.Ipv6Addresses[0].Ipv6Address)
```
Hey @mtulio That's right, but the primary IPv6 requirement actually depends on the [target type](https://docs.aws.amazon.com/elasticloadbalancing/latest/network/load-balancer-target-groups.html#target-type) of the Target Group. In our case, the following type is applicable:

- `instance`: The instance must be in the same VPC as NLB and is assigned a "stable" IPv6 address. Unlike IPv4, IPv6 address, by default, can be re-assigned elsewhere. Thus, we need to set `IsPrimaryIpv6: true` to make sure the IPv6 address is tied to the ENI of the instance for its entire lifecycle. I guess "primary" on AWS means "stable/owned".
  > When registering targets by instance ID for a IPv6 target group, the targets must have an assigned primary IPv6 address. To learn more, see [IPv6 addresses](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-instance-addressing.html#ipv6-addressing) in the Amazon EC2 User Guide

- `IP`: There is no requirement for a "stable" IPv6 address (or setting flag `IsPrimaryIpv6`). But the IPv6 address must be within the VPC IPv6 CIDR of the cluster.
  > When registering targets by IP address for an IPv6 target group, the IP addresses that you register must be within the VPC IPv6 CIDR block or within the IPv6 CIDR block of a peered VPC.

My understanding is that we should prefer `instanceID` target type in order to ensure traffic will always reach the intended target (i.e. cluster nodes). Registering with IP address will pin the traffic to the IP, which may not be the cluster node at all as it can be re-assigned to other services, for example, another non-cluster EC2 instance.

@nrb IIUC, we should make the CCM prefer registering by `instance` when, let's say, primary IPv6 is available. Otherwise, fall back to register by the IP address?

sadasu commented on pkg/providers/v1/aws_loadbalancer.go:
```diff
@@ -784,6 +787,16 @@ func (c *Cloud) ensureTargetGroup(ctx context.Context, targetGroup *elbv2types.T
 
+		// Set IP address type based on annotation
+		if tgIPAddressType, ok := annotations[ServiceAnnotationLoadBalancerTargetGroupIPAddressType]; ok {
+			if tgIPAddressType == string(elbv2types.TargetGroupIpAddressTypeEnumIpv6) || tgIPAddressType == string(elbv2types.TargetGroupIpAddressTypeEnumIpv4) {
```
Can we document why we can have only ipv4 or ipv6 and not say [ipv4,ipv6] for this annotation?

sadasu commented on pkg/providers/v1/aws_loadbalancer.go:
```diff
@@ -1111,23 +1112,49 @@ func (c *Cloud) updateInstanceSecurityGroupsForNLB(ctx context.Context, lbName s
 		for _, cidr := range cidrs {
-			desiredPerms.Insert(ec2types.IpPermission{
+			perm := ec2types.IpPermission{
 				IpProtocol: aws.String(protocol),
```
Should IpProtocol be set below based on result of isIPv6CIDR(cidr)?

nrb replied:
Ah, yes, we should support `Ipv6Ranges`, will update.

sadasu commented on pkg/providers/v1/aws_loadbalancer.go:
```diff
@@ -94,6 +95,71 @@ func isLBExternal(annotations map[string]string) bool {
+	}
+
+	// Fall back to spec.ipFamilies and spec.ipFamilyPolicy
+	if service.Spec.IPFamilyPolicy != nil {
```
Also, do we have to make sure that when annotation and spec fields are both present, they match?

sadasu reviewed (commented):
@nrb Thanks for making these updates. A few comments inline.

nrb commented:
govulncheck notice will be resolved by pulling https://github.com/kubernetes/kubernetes/pull/136820 in during a kube bumpl.

tthvo commented on pkg/providers/v1/aws_loadbalancer.go:
I guess we need to configure the `IpAddressType` according to service `ipFamilies`, right? 

Maybe:

```go
func getLoadBalancerIPAddressTypeFromService(service *v1.Service) elbv2types.IpAddressType {
    if serviceRequestsIPv6(service) {
        return elbv2types.IpAddressTypeDualstack
    }
    return elbv2types.IpAddressTypeIpv4
}
```

kmala commented on pkg/providers/v1/aws.go:
```diff
@@ -2452,13 +2505,14 @@ func (c *Cloud) EnsureLoadBalancer(ctx context.Context, clusterName string, apiS
 			internalELB,
 			annotations,
 			securityGroups,
+			apiService,
```
nit: if we are passing the entire service then we can skip passing other params like service name, annotation right?

tthvo commented on docs/service_controller.md:
```diff
@@ -85,4 +85,66 @@ metadata:
+    type: LoadBalancer
+    ipFamilies:
+        - IPv6
+        - IPv4
```
nit: The list indentation is a lot more than usual. Let's format it a bit 😁

tthvo reviewed (commented):
I just have a few questions 🙏  But overall, I think we are having great progress!

nrb commented on pkg/providers/v1/aws_loadbalancer.go:
```diff
@@ -159,6 +183,16 @@ func (c *Cloud) ensureLoadBalancerv2(ctx context.Context, namespacedName types.N
+		if serviceRequestsIPv6(service) {
+			return nil, fmt.Errorf("single stack IPv6 is not supported for network load balancers")
+		}
+	}
```
Yep, makes sense.

nrb commented on pkg/providers/v1/aws.go:
```diff
@@ -3221,6 +3275,11 @@ func (c *Cloud) buildSecurityGroupsToDelete(ctx context.Context, service *v1.Ser
 // EnsureLoadBalancerDeleted implements LoadBalancer.EnsureLoadBalancerDeleted.
 func (c *Cloud) EnsureLoadBalancerDeleted(ctx context.Context, clusterName string, service *v1.Service) error {
+	// If spec.loadBalancerClass is set, another controller is responsible for this Service
+	if service.Spec.LoadBalancerClass != nil {
```
Good catch on the `wantsLoadBalancer` function, wasn't aware of that.

@kmala this was mostly to be explicit about the contract around load balancer classes, but Thuan's pointed out the service controller handling that, so I'll remove this check.

tthvo commented on pkg/providers/v1/aws_loadbalancer.go:
```diff
@@ -94,6 +95,29 @@ func isLBExternal(annotations map[string]string) bool {
+// getTargetGroupIPAddressTypeFromService determines the IP address type for the target group
+// based on the Service's spec.ipFamilies field. According to Kubernetes dual-stack documentation:
+// - If ipFamilies[0] is IPv6, the target group should use IPv6
+// - If ipFamilies is not set or ipFamilies[0] is IPv4, the target group should use IPv4 (default)
```
> I think this code might be wrong if the user provides an IPv6 CIDR to --service-cluster-ip-range and then leaves spec.IPFamilies unset.

Right... I am not sure how we can handle it without knowing the configured service CIDRs of the cluster :disappointed: My hope is that https://github.com/kubernetes/cloud-provider-aws/pull/1313#discussion_r2856071476 is true. From my local testing with [`kinD`](https://kind.sigs.k8s.io/docs/user/configuration/#networking), the `ipFamilies` are always non-empty after creation :eyes: 

If not, I found CCM azure provider uses the same logic as yours: https://github.com/kubernetes-sigs/cloud-provider-azure/blob/b0192b1b05f449d8b008a4b1519ff9324d527c1f/pkg/provider/azure_utils.go#L302-L315

damdo commented:
/test pull-cloud-provider-aws-e2e-kubetest2-quick

nrb committed:
Add dual-stack e2e test cases

Signed-off-by: Nolan Brubaker <nolan@nbrubaker.com>

nrb commented:
Current e2e failure is legit - the provisioned cluster is not set up for IPv6.

```
   STEP: creating LoadBalancer service in Kubernetes @ 03/03/26 17:45:37.338
  I0303 17:45:37.386843 32011 loadbalancer.go:508] Unexpected error: 
      <*errors.errorString | 0xc000321140>: 
      failed to create LoadBalancer Service "lbconfig-test-clb-ds": Service "lbconfig-test-clb-ds" is invalid: spec.ipFamilies[1]: Invalid value: "IPv6": not configured on this cluster
      {
          s: "failed to create LoadBalancer Service \"lbconfig-test-clb-ds\": Service \"lbconfig-test-clb-ds\" is invalid: spec.ipFamilies[1]: Invalid value: \"IPv6\": not configured on this cluster",
      }
  [FAILED] in [It] - /home/prow/go/src/k8s.io/cloud-provider-aws/tests/e2e/loadbalancer.go:508 @ 03/03/26 17:45:37.387
  STEP: Destroying namespace "cloud-provider-aws-593" for this suite. @ 03/03/26 17:45:37.387
  << Timeline 
```

Will work on getting support for that plumbed in to kubetest-ec2.

nrb committed:
Add support for dual networking stack services

- Create load balancers according to the Kubernetes Service API

Signed-off-by: Nolan Brubaker <nolan@nbrubaker.com>
Assisted-by: Claude Sonnet 4.5 <noreply@anthropic.com>

nrb committed:
Support IP address type changes on update

Signed-off-by: Nolan Brubaker <nolan@nbrubaker.com>

nrb committed:
Hash target group names with IP address family

Changing the IP address type would invalidate the target group name,
because you cannot have an IPv4 and IPv6 target for the same
port/protocol set.

Signed-off-by: Nolan Brubaker <nolan@nbrubaker.com>

nrb committed:
Take IPv6 CIDRs into account for all open traffic

Signed-off-by: Nolan Brubaker <nolan@nbrubaker.com>

nrb committed:
fixup! Support IP address type changes on update

Signed-off-by: Nolan Brubaker <nolan@nbrubaker.com>

nrb committed:
Gracefully fall back to IPv4 for CLBs if able

Signed-off-by: Nolan Brubaker <nolan@nbrubaker.com>
